File: test/shire/slug_test.exs
defmodule Shire.SlugTest do
  use ExUnit.Case, async: true

  alias Shire.Slug

  describe "valid?/1" do
    test "accepts simple lowercase name" do
      assert Slug.valid?("hello")
    end

    test "accepts name with numbers" do
      assert Slug.valid?("agent1")
    end

    test "accepts name with dashes" do
      assert Slug.valid?("my-agent")
    end

    test "accepts single character" do
      assert Slug.valid?("a")
    end

    test "accepts single digit" do
      assert Slug.valid?("1")
    end

    test "accepts name with multiple dashes" do
      assert Slug.valid?("my-cool-agent")
    end

    test "rejects uppercase letters" do
      refute Slug.valid?("MyAgent")
    end

    test "rejects spaces" do
      refute Slug.valid?("my agent")
    end

    test "rejects special characters" do
      refute Slug.valid?("my_agent")
      refute Slug.valid?("my.agent")
      refute Slug.valid?("my@agent")
    end

    test "rejects leading dash" do
      refute Slug.valid?("-agent")
    end

    test "rejects trailing dash" do
      refute Slug.valid?("agent-")
    end

    test "rejects empty string" do
      refute Slug.valid?("")
    end

    test "rejects nil" do
      refute Slug.valid?(nil)
    end

    test "rejects name longer than 63 characters" do
      refute Slug.valid?(String.duplicate("a", 64))
    end

    test "accepts name with exactly 63 characters" do
      assert Slug.valid?(String.duplicate("a", 63))
    end
  end

  describe "slugify/1" do
    test "lowercases the string" do
      assert Slug.slugify("MyAgent") == "myagent"
    end

    test "replaces spaces with dashes" do
      assert Slug.slugify("my agent") == "my-agent"
    end

    test "replaces underscores with dashes" do
      assert Slug.slugify("my_agent") == "my-agent"
    end

    test "replaces special characters with dashes" do
      assert Slug.slugify("my@agent") == "my-agent"
    end

    test "collapses consecutive dashes" do
      assert Slug.slugify("my--agent") == "my-agent"
    end

    test "strips leading and trailing dashes" do
      assert Slug.slugify("-my-agent-") == "my-agent"
    end

    test "handles complex case" do
      assert Slug.slugify("  My Cool Agent! ") == "my-cool-agent"
    end

    test "result is a valid slug" do
      assert Slug.valid?(Slug.slugify("Hello World 123"))
    end
  end
end

File: src/services/slug.test.ts
import { describe, it, expect } from "bun:test";
import { valid, slugify } from "./slug";

describe("valid", () => {
  it("accepts simple lowercase name", () => {
    expect(valid("hello")).toBe(true);
  });

  it("accepts name with numbers", () => {
    expect(valid("agent1")).toBe(true);
  });

  it("accepts name with dashes", () => {
    expect(valid("my-agent")).toBe(true);
  });

  it("accepts single character", () => {
    expect(valid("a")).toBe(true);
  });

  it("accepts single digit", () => {
    expect(valid("1")).toBe(true);
  });

  it("accepts name with multiple dashes", () => {
    expect(valid("my-cool-agent")).toBe(true);
  });

  it("rejects uppercase letters", () => {
    expect(valid("MyAgent")).toBe(false);
  });

  it("rejects spaces", () => {
    expect(valid("my agent")).toBe(false);
  });

  it("rejects special characters", () => {
    expect(valid("my_agent")).toBe(false);
    expect(valid("my.agent")).toBe(false);
    expect(valid("my@agent")).toBe(false);
  });

  it("rejects leading dash", () => {
    expect(valid("-agent")).toBe(false);
  });

  it("rejects trailing dash", () => {
    expect(valid("agent-")).toBe(false);
  });

  it("rejects empty string", () => {
    expect(valid("")).toBe(false);
  });

  it("rejects nil/undefined", () => {
    expect(valid(null)).toBe(false);
    expect(valid(undefined)).toBe(false);
  });

  it("rejects name longer than 63 characters", () => {
    expect(valid("a".repeat(64))).toBe(false);
  });

  it("accepts name with exactly 63 characters", () => {
    expect(valid("a".repeat(63))).toBe(true);
  });
});

describe("slugify", () => {
  it("lowercases the string", () => {
    expect(slugify("MyAgent")).toBe("myagent");
  });

  it("replaces spaces with dashes", () => {
    expect(slugify("my agent")).toBe("my-agent");
  });

  it("replaces underscores with dashes", () => {
    expect(slugify("my_agent")).toBe("my-agent");
  });

  it("replaces special characters with dashes", () => {
    expect(slugify("my@agent")).toBe("my-agent");
  });

  it("collapses consecutive dashes", () => {
    expect(slugify("my--agent")).toBe("my-agent");
  });

  it("strips leading and trailing dashes", () => {
    expect(slugify("-my-agent-")).toBe("my-agent");
  });

  it("handles complex case", () => {
    expect(slugify("  My Cool Agent! ")).toBe("my-cool-agent");
  });

  it("result is a valid slug", () => {
    expect(valid(slugify("Hello World 123"))).toBe(true);
  });
});
